### PR TITLE
Ribasim Python: Add-API

### LINKED ISSUE #252: Ribasim Python: stop exposing `Ribasim.Node` and node IDs to users
Thanks to @SouthEndMusic's work, users will soon have fewer chances to mess up their model input when using Ribasim Python.
https://github.com/Deltares/Ribasim/issues/245

What would be even better, if users wouldn't have to think about node IDs at all. They would create specific node types with the corresponding geometries attached. When new nodes are created that way, they would access a singleton allocator which returns new and unique node IDs.

For creating edges, the interfaces might change like this:

```diff
- from_id = np.array(
-    [1, 2], dtype=np.int64
-)
- to_id = np.array(
-    [2, 3], dtype=np.int64
-)

+ from_id = [basin.nodes[0], linear_resistance.nodes[0]]
+ to_id = [linear_resisitance.nodes[0], basin.nodes[1]]
```

### LINKED ISSUE #760: Implement object based "add" API in Python
Will be longer than sprint 10, probably done in 12.

### PR DESCRIPTION
**Pull request by Hofer-Julian**
Fixes #760
Fixes #252

Follow ups:
- More validation
- Avoid blocking of database by qgis
- Plotting of control edges
- Reading of model

### CODE REVIEW

**Comment by visr** [comment]
For updating `ribasim_testmodels`, I'm not including the `crs="EPSG:28992"` anymore. It is not important and I don't think it is supported yet anyway. Also I'm leaving out the `# Setup X` comments, they are quite obvious from the code, especially with this PR. I decided to first create all nodes and then all edges rather than having that interwoven.

File: python/ribasim/ribasim/model.py [hunk 1]
```diff
@@ -456,92 +221,27 @@ def plot(self, ax=None, indicate_subnetworks: bool = True) -> Any:
...
-                for var in static.columns:
-                    if var not in ["node_id", "control_state"]:
-                        value = row[var]
-                        if value is not None:
-                            names_and_values.append(f"{var} = {value}")
-
-                out += ", ".join(names_and_values) + "\n"
-
-            print(out)
+        raise NotImplementedError()
```
Comment (SouthEndMusic) [inline python/ribasim/ribasim/model.py:247]: (As the person who added it) I wonder whether this is worth maintaining

Comment (Hofer-Julian) [reply python/ribasim/ribasim/model.py:247]: Yay to deleting code

**Review by visr — COMMENTED** [review]
Left some comments. I see that models without edges (bucket) now don't have an edge table. Perhaps it would be better to then write an empty edge table?

File: python/ribasim/ribasim/geometry/edge.py [hunk 2]
```diff
@@ -41,6 +49,39 @@ class Edge(SpatialTableModel[EdgeSchema]):
...
+                "from_node_type": pd.Series([from_node.node_type], dtype=str),
+                "from_node_id": pd.Series([from_node.node_id], dtype=int),
+                "to_node_type": pd.Series([to_node.node_type], dtype=str),
+                "to_node_id": pd.Series([to_node.node_id], dtype=int),
+                "edge_type": pd.Series([edge_type], dtype=str),
+                "name": pd.Series([name], dtype=str),
+                "subnetwork_id": pd.Series([subnetwork_id], dtype=pd.Int64Dtype()),
+            },
+            geometry=geometry_to_append,
+            crs="EPSG:28992",
```
[the remark this answers is not in the document]
Comment (Hofer-Julian) [reply python/ribasim/ribasim/geometry/edge.py:77]: I removed it, this leads to the following warning though:
```
UserWarning: 'crs' was not provided.  The output dataset will not have projection information defined and may not be usable in other systems.
```

[the remark this answers is not in the document]
Comment (visr) [reply python/ribasim/ribasim/geometry/edge.py]: Yeah that's fine for now.

File: .vscode/settings.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
     "[julia]": {
         "editor.formatOnSave": true
     },
-    "notebook.formatOnSave.enabled": true,
+    "notebook.formatOnSave.enabled": false,
```
[the remark this answers is not in the document]
Comment (Hofer-Julian) [reply .vscode/settings.json]: No, it is really broken

**Review by visr — APPROVED** [review]

**Pull request merged**